File: linkedlist.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None
        
    def __repr__(self):
        return str(f" Node: {self.data}")
    
    

node1 = Node('1st Node')
node2 = Node('2nd Node')
node1.next = node2
node2.next = Node('3rd Node')

def print_all_nodes(node: Node):
    current = node
    while current:
        print(current)
        current = current.next

#print_all_nodes(node1)

class LinkedList:
    def __init__(self):
        self.head = None
    
    def __contains__(self,data):
        current = self.head
        while current:
            if  current.data == data:
                return True
            current = current.next
        return False
            
    def __repr__(self):
        current = self.head
        new_str = ''
        new_str += str(current.data)
        current = current.next
        while current:
            new_str += '->'
            new_str += str(current.data)
            current  = current.next
        return new_str
    
    def __iter__(self):
        current = self.head
        while current:
            if current.data % 2 == 0:
                yield current
            current = current.next
        
        
    def __len__(self):
        count = 0
        current = self.head
        while current:
            count += 1
            current = current.next
        return count
        
    def append(self,data):
        node = Node(data)
        node.next = self.head
        self.head = node
        
    def delete(self,data):
        current = self.head
        if current.data == data:
            self.head = self.head.next
            return current
        while current.next:
            if current.next.data == data:
                deleted = current.next
                current.next = current.next.next
                return deleted
            current = current.next
        return None

    ######
    def isCycle(node):
    marker1 = node
    marker2 = node
    # print (node.value)
    while marker2 != None and marker2.nextnode != None:
        marker1 = marker1.nextnode
        marker2 = marker2.nextnode.nextnode
        if marker2 == marker1:
            return True

    
    return False
    
    ######


    def reverse(node):
        current = node
        prev = None
        nextNode = None

        while current:
            nextNode = current.nextnode
            current.nextnode = prev
            prev = current
            current = nextNode
        return prev

    #####
    def merge(head1: Node, head2: Node):
    if head1 is None:
        return head2
    if head2 is None:
        return head1
        
    tail = Node()
    while not(head1 is None or head2 is None):
        if head1.data < head2.data:
            current = head1
            head1 = head1.next
        else:
            current = head2
            head2 = head2.next
        tail.next = current
        tail = tail.next
    tail.next = head1 or head2
    return tail.next
    ######
    def remove_duplicates(head: Node)-> list:
    current = head
    prev = None
    dic = {}
    
    while current:
        if current.data in dic:
            
            prev.next = current.next
            current = None
        else:
            
            dic[current.data] = 1
            prev = current
        current = prev.next

File: linkedList.js
// Question 1: Implement a Linked List

class Node {
    constructor(data = null){
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null
    }
    
    appendToTail(data){ // add new Node with data to tail
        let tail = this.head;
        if (!tail) {
          this.head = new Node(data);
        } else {
          while (tail.next !== null) {
            tail = tail.next;
          }
          tail.next = (new Node(data));
        }  
    }

    prependToHead(data){ // add new Node with data to head
        const newHead = new Node(data);
        const currentHead = this.head;
        this.head = newHead;
        if (currentHead) {
          this.head.next = currentHead;
        }
    }
    
    removeHead() { // removed the first Node in the LinkedList and returns its data
        const removedHead = this.head;
        if (!removedHead) {
          return;
        }
        this.head = removedHead.next;
        return removedHead.data;
    }

    contains(data){ // returns true is any Node in the LinkedList contains data, false otherwise
        let currentNode = this.head
        while(currentNode) {
            if(currentNode.data === data){
                return true
            }
            currentNode = currentNode.next
        }
        return false
    }

    length(){ //returns the length of the LinkedList as an integer value
        let currentNode = this.head
        let size = 0;
        while(currentNode) {
            size ++
            currentNode = currentNode.next
        }
        return size
    }
}

// Question 2: Cycle Check
const isCyclic = (headNode) => { //returns a boolean
    let node1 = headNode
    let node2 = headNode.next
    while(node1 && node2) {
        if(node1 === node2) {
            return true
        }
        node1 = node1.next;
        if(node2.next){
            node2 = node2.next.next;
        } else {
            return false;
        }
    }
    return false;
};

// Question 3: Reverse a Linked List
const reverse = (headNode) => { //returns the new headNode
    var node = headNode;
    var previous = null;

    while(node) {
        // save next or you lose it!!!
        var save = node.next;
        // reverse pointer
        node.next = previous;
        // increment previous to current node
        previous = node;
        // increment node to next node or null at end of list
        node = save;
    }
    return previous;   // Change the list head !!!
};

// Question 4: Merge Two Lists
const mergeLists = (head1, head2) => { //returns the headNode
    let node = head1;
    while(node) {
        if(!node.next) {
            node.next = head2;
            break;
        }
        node = node.next
    }
    return head1
};

// Question 5: Remove duplicates
const removeDuplicates = (headNode) => { //returns the headNode
    let obj = {};
    let currentNode = headNode;
    let prev = null;
    while(currentNode){
        if(currentNode.data in obj){
            prev.next = currentNode.next
        } else {
            obj[currentNode.data] = true;
        }
        prev = currentNode
        currentNode = currentNode.next
    }
    return headNode
};

module.exports = {
    Node, LinkedList, isCyclic, reverse, mergeLists, removeDuplicates
}